Add new Overdrive features

bpalme opened the pull request:
OK here are the changes to satisfy the two issues referenced in the title and bring the Overdrive integration back into working order. One things to note... I went a different direction on the core template. Previously we had some of the holdings information in the core which was somewhat clumsy to make it work by setting some variable with HTML comment and then calling the parent core tab. I elected to make it more like other content and moved all of the holding info to a holdings tab. I think that will make more sense and make the integration more uniform. If you are OK with that, then it might make sense to remove the call in the main template to show the contents of those variables. I hope I didn't lose all the formatting fixes that were done in the intervening time. I tried to fold those in, but I doubt I got all of them.

TODO
- [x] Resolve [VUFIND-1499](https://openlibraryfoundation.atlassian.net/browse/VUFIND-1499) and [VUFIND-1375](https://openlibraryfoundation.atlassian.net/browse/VUFIND-1375) when merging.

Commit:
update overdrive integration for latest API changes

Commit:
update overdrive integration for latest API changes

Commit:
redo the formatting and fixes from VF since our versions diverged

Commit:
redo the formatting and fixes from VF since our versions diverged

bpalme committed:
Merge branch 'vufind-org:dev' into VUFIND-1499+VUFIND-1375

Commit:
cleaning up errors from phpcs

Commit:
cleaning up errors from phpcs

Commit:
more formatting fixes

Commit:
more formatting fixes

Commit:
more formatting fixes

Commit:
added some corresponding test functions

xmorave2 committed:
Fix return types in Aleph driver (#2943)

xmorave2 committed:
Fix page param out of range at search results (#2940)

EreMaijala committed:
Avoid division by zero when result limit is 0. (#2945)

EreMaijala committed:
Replace strtolower with mb_strtolower to avoid encoding issues. (#2946)

ckaz committed:
Corrects SCSS compile extend error (#2922)

demiankatz committed:
Update SCSS Bootstrap to match Less version. (#2948)

demiankatz committed:
Convert translation maps to UTF-8. (#2942)

demiankatz committed:
Improve SCSS-related CI behavior. (#2924)

crhallberg committed:
autocomplete.js accessibility patch. (#2949)

- Resolves VUFIND-1616.

demiankatz committed:
Bump version to 9.0.2.

ThoWagen committed:
moved ExplodeSettingTrait to Feature subdir (#2951)

demiankatz committed:
Update outdated/incorrect EDS.ini language. (#2950)

maccabeelevine committed:
Add EBSCO Publication Finder data source (#2944)

maccabeelevine committed:
EPF: Test record driver and changes to EDS Backend (#2953)

demiankatz committed:
Fix copy-and-paste errors in comments. (#2954)

maccabeelevine committed:
[VUFIND-1620] Remove inactive EDS facet_limit (#2955)

xmorave2 committed:
Add third cache server for obalkyknih.cz (#2956)

LuomaJuha committed:
Add Northern Sámi language (#2957)

demiankatz committed:
Add support for Laminas 'Dumb' CAPTCHA. (#2959)

demiankatz committed:
Remove unused use statements from Util Commands. (#2960)

maccabeelevine committed:
LibGuidesAZ: Add results recommendation module (#2964)

demiankatz committed:
php-cs-fixer: Remove disabled braces rule configuration. (#2963)

xrosecky committed:
active favorites in menu

mtrojan-ub committed:
Browse: Use same localized number formats like in facets (#2968)

demiankatz committed:
Rename abstract SearchObject class to AbstractSearchObject.

demiankatz committed:
Add deprecated SearchObject for back-compatibility.

maccabeelevine committed:
Refactor *Deferred recommendation modules (#2966)

maccabeelevine committed:
LibGuides: Add results recommendation module (#2972)

demiankatz committed:
[VUFIND-1622] Fix EDS date range facet display. (#2970)

maccabeelevine committed:
EPF: Add results recommendation module (#2973)

ckaz committed:
Turn my account menu into ul, because it is a list (#2914)

demiankatz committed:
Fix minor style inconsistency.

demiankatz committed:
Revert "Fix minor style inconsistency."

This reverts commit 21dc050cef3d6a78e95e59919b7866a29b533e68.

elsenhans committed:
Wrap h2 around title and author, add span around author (#2958)

maccabeelevine committed:
EPF: Add holdings list to record view (#2975)

Also escape text within full-text-links on the results page.

demiankatz committed:
Fix broken test.

demiankatz committed:
Refactor account view record details to subtemplate. (#2976)

demiankatz committed:
[VUFIND-1585] Add Mink tests for library cards. (#2971)

dependabot[bot] committed:
Bump league/oauth2-server from 8.5.1 to 8.5.3 (#2980)

Bumps [league/oauth2-server](https://github.com/thephpleague/oauth2-server) from 8.5.1 to 8.5.3.
- [Release notes](https://github.com/thephpleague/oauth2-server/releases)
- [Changelog](https://github.com/thephpleague/oauth2-server/blob/master/CHANGELOG.md)
- [Commits](https://github.com/thephpleague/oauth2-server/compare/8.5.1...8.5.3)

---
updated-dependencies:
- dependency-name: league/oauth2-server
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

RLangeUni committed:
Handle OR operator for SortFacetList (#2981)

maccabeelevine committed:
Add recommendation module for LibGuides Profiles (#2977)

demiankatz committed:
Remove confusing with_selected string. (#2978)

maccabeelevine committed:
Test LibGuides connector and LibGuidesProfile (#2984)

damien-git committed:
FOLIO: Remove empty holdings statements (#2983)

demiankatz committed:
Shibboleth: clean up comments with use statement. (#2986)

demiankatz committed:
Add simulated single sign-on for testing purposes. (#2987)

demiankatz committed:
Whitespace/capitalization cleanup in comments and messages. (#2988)

- Also removes a few obsolete/inaccurate Overdrive-related comments.

demiankatz committed:
Fix TranslatorAwareTrait comments. (#2990)

demiankatz committed:
Add missing container dependency to ChannelsController. (#2989)

demiankatz committed:
Fix advanced search filter operator handling. (#2992)

- Resolves VUFIND-1623.

ThoWagen committed:
Inheritance of local dirs for configuration files (#2947)

mtrojan-ub committed:
Use CachingDownloader for Google and Orb covers (#2659)

- Includes new mechanism for disabling caching
- Includes new configuration settings for service-specific cache settings
---------

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

crhallberg committed:
Cleanup: remove outdated and unused CSS file. (#2997)

crhallberg committed:
fix: jstree elements overflow text properly with ellipses. (#2996)

- Resolves VUFIND-1621.

demiankatz committed:
Bump version number.

damien-git committed:
Optimized batch_import_marc (#2995)

- Allows multiple files to be passed to import-marc.sh (Linux + Windows)
- Utilizes multiple file loading to improve performance of batch-import-marc.sh (Linux only)

---------

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

Commit:
multiple changes for overdrive API changes

Commit:
multiple changes for overdrive API changes

Commit:
fixing other small problems plus recompiled less

Commit:
fix language entry out of order

Commit:
change supportsOpenURL back to false

Commit:
formatting fixes

Commit:
more formatting fixes

Commit:
removed a debugged comment that got missed

demiankatz commented:
@bpalme, since development work on release 9.1 will be wrapping up soon, I'm moving this to the 10.0 milestone. If you get the work done very soon, I'm still open to merging it into the earlier release, but since I imagine it may take a while longer to get all the details right, I think 10.0 is probably the most realistic version target. :-)

bpalme commented:
Ok I think  that makes sense.

From: Demian Katz ***@***.***>
Sent: Thursday, August 31, 2023 5:45 AM
To: vufind-org/vufind ***@***.***>
Cc: Brent Palmer ***@***.***>; Mention ***@***.***>
Subject: Re: [vufind-org/vufind] Add new Overdrive features (PR #3025)


@bpalme<https://github.com/bpalme>, since development work on release 9.1 will be wrapping up soon, I'm moving this to the 10.0 milestone. If you get the work done very soon, I'm still open to merging it into the earlier release, but since I imagine it may take a while longer to get all the details right, I think 10.0 is probably the most realistic version target. :-)

—
Reply to this email directly, view it on GitHub<https://github.com/vufind-org/vufind/pull/3025#issuecomment-1700801154>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABC6W3Q3QMSBCBVK54AQUM3XYBTJ5ANCNFSM6AAAAAA3FOVUKA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

demiankatz commented:
@bpalme, just a quick update to say that I've marked this PR "on hold" until you are ready, since there is nothing to review until you are able to share your latest work. No rush, but let me know when you'd like me to remove that label.

Commit:
added getURLs fxn to solroverdrive rm

demiankatz committed:
Merge branch 'dev' into VUFIND-1499+VUFIND-1375

demiankatz committed:
php-cs-fixer

demiankatz committed:
Delete obsolete file.

demiankatz committed:
Fix less.

demiankatz committed:
Revert bad changes.

demiankatz commented:
@bpalme, thanks for making some progress here! I just took some time to resolve all the conflicts and get this properly based against the dev branch. I hope I didn't break anything in the process! Could you please review this when you get a chance and correct anything that I may have messed up? Once you're happy that this is complete and functional, I can give it a closer review from my end.

demiankatz commented:
@bpalme, I notice that there is some kind of parsing error in `themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml` -- any idea what that's about?

bpalme commented:
Thank you so much. I'll try to make some progress on it soon.

Brent

From: Demian Katz ***@***.***>
Sent: Thursday, November 9, 2023 8:30 AM
To: vufind-org/vufind ***@***.***>
Cc: Brent Palmer ***@***.***>; Mention ***@***.***>
Subject: Re: [vufind-org/vufind] Add new Overdrive features (PR #3025)


@bpalme<https://github.com/bpalme>, thanks for making some progress here! I just took some time to resolve all the conflicts and get this properly based against the dev branch. I hope I didn't break anything in the process! Could you please review this when you get a chance and correct anything that I may have messed up? Once you're happy that this is complete and functional, I can give it a closer review from my end.

—
Reply to this email directly, view it on GitHub<https://github.com/vufind-org/vufind/pull/3025#issuecomment-1803939498>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABC6W3UDHPDFIMGSAPVTAZTYDTSGXAVCNFSM6AAAAAA3FOVUKCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMBTHEZTSNBZHA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Commit:
fix parsing error in template

Commit:
fix Fake connector to match method signature change

Commit:
fix typo in class name

Commit:
phpcs fixes

demiankatz commented on config/vufind/Overdrive.ini:
```diff
@@ -36,6 +36,10 @@ noAccessString = ""
 ; more overhead with it on. Default is false.
 consortiumSupport = false
 
+;AJAX Status is supported, but you can turn it off it's slowing the search results
```
I think this comment is missing an "if" between "off" and "it's." We also by convention include spaces after the `;` in comments.

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php:
```diff
@@ -250,8 +251,9 @@ public function getPreviewLinks()
-        // TODO: add this as an Overdrive configuration to turn it off
-        return true;
+        $this->debug('ajax status:' . $this->config->enableAjaxStatus);
+        $this->debug('all configs:' . print_r($this->config, true));
```
Do we need all these debug outputs here?

bpalme replied:
No. Should I generally go through all the comments and trim it down to the most useful?

demiankatz replied:
It wouldn't hurt. As a rule of thumb: if it's going to help somebody who's having trouble with Overdrive, feel free to leave it. If it's something you needed during development but no longer expect to require, get rid of it. :-)

bpalme replied:
Remove quite a few debug statements

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php:
```diff
@@ -331,24 +333,34 @@ public function getOverdriveAvailability()
+            $result->data = [];
             foreach ($checkouts as $checkout) {
-                if (strtolower($checkout->reserveId) == $overdriveID) {
+                $this->debug("comparing: {$checkout->reserveId} to $overdriveID");
```
This debug message seems potentially outdated, since you're doing something more nuanced now.

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php:
```diff
@@ -563,4 +575,41 @@ public function getRealTimeTitleHold()
+    {
+        $permlink =  $this->getPermanentLink();
+        //todo translate
+        $desc = 'Overdrive Resource Page';
```
Record drivers have access to the translator, so this TODO is not hard at all -- just make sure the string is in the language file, and change this to `$desc = $this->translate(...)`. Additionally, if you're hard-coding $desc, there's no need to do the `?: $permlink` in the next line, since $desc will never be false.

demiankatz commented on module/VuFind/src/VuFind/RecordTab/OverdriveHoldingsFactory.php:
```diff
@@ -0,0 +1,62 @@
+     */
+    public function __invoke(\Interop\Container\ContainerInterface $container, $requestedName, array $options = null)
+    {
+        return new \VuFind\RecordTab\OverdriveHoldings();
```
If you're just creating the object with no extra parameters, there is no need to build a whole factory. You can just configure the plugin manager to use `\Laminas\ServiceManager\Factory\InvokableFactory` to build the object. And speaking of that, I don't see any changes to register this new class in the PluginManager. I wonder if I accidentally lost those changes when resolving conflicts. Please let me know if you need help getting them back!

bpalme replied:
Hmm. I don't think I ever put it in the PluginManager but it seems to be working without it listed. But I added it and removed the unnecessary factory and that works too.

demiankatz replied:
Thanks! The plugin managers have some default behavior for building objects when configuration is missing, so that's probably why this worked before... but I prefer to be explicit about configuration when possible, so thanks for adding the missing details!

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/FakeOverdriveConnector.php:
```diff
@@ -220,31 +298,50 @@ public function returnResource($resourceID)
 
     /**
-     * Lock In Overdrive Resource for a particular format
+     * Returns
```
This description doesn't seem right for some reason.

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -873,12 +1057,42 @@ public function getMetadata($overDriveIds = [])
+     *
+     * @return array inital array with results of metadata attached as "metadata" property
+     *
+     * @todo if more tan 25 passed in, make multiple calls
```
Typo "tan" for "than"

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/FakeOverdriveConnector.php:
```diff
@@ -253,12 +350,30 @@ public function lockinResource($overDriveId, $format)
      *
      * @return array results of metadata fetch
+     *
+     * @todo if more tan 25 passed in, make multiple calls
```
Do we need this todo in the fake version?

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/FakeOverdriveConnector.php:
```diff
@@ -253,12 +350,30 @@ public function lockinResource($overDriveId, $format)
+     *
+     * @return array inital array with results of metadata attached as "metadata" property
+     *
+     * @todo if more tan 25 passed in, make multiple calls
```
Do we need this todo in the fake version?

demiankatz replied:
These todos are still here, though my comments about them are resolved. Should they be removed?

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/FakeOverdriveConnector.php:
```diff
@@ -253,12 +350,30 @@ public function lockinResource($overDriveId, $format)
+     * @param array $overDriveTitles Assoc array of objects with OD IDs as keys (generally what
+     *                               you get from getCheckouts and getHolds)
+     *
+     * @return array inital array with results of metadata attached as "metadata" property
```
Typo: inital

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -873,12 +1057,42 @@ public function getMetadata($overDriveIds = [])
+     * @param array $overDriveTitles Assoc array of objects with OD IDs as keys (generally what
+     *                               you get from getCheckouts and getHolds)
+     *
+     * @return array inital array with results of metadata attached as "metadata" property
```
Typo: inital

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -843,9 +953,82 @@ public function getFormatNames()
+                $checkoutResult = $this->getCheckouts();
+                $checkoutData = $checkoutResult->data;
+                foreach ($result->data->products as $key => $issue) {
+                    if (isset($checkoutData[strtolower($issue->id)])) {
```
This if/else could be simplified to:

```
$issue->checkedout = isset($checkoutData[strtolower($issue->id)]);
```

demiankatz reviewed (changes requested):
Thanks for the progress here, @bpalme! There's a lot of code to review, so I haven't looked thoroughly at everything yet, but I thought I'd get the ball rolling with a few comments based on an initial quick scan of what has changed. See below!

Please let me know if/when you'd like help with hands-on testing as well! Obviously my ability to test is limited since I do not have access to OverDrive, but with the fake connector, I can at least do some local testing. I just don't want to start that before you think the code is ready to run, so please keep me informed of the status. :-)

demiankatz committed:
Merge branch 'dev' into VUFIND-1499+VUFIND-1375

demiankatz commented:
@bpalme, I also resolved the latest round of conflicts while I was looking at this anyway. (Nothing complicated -- just had to recompile the CSS).

bpalme commented:
> Thanks for the progress here, @bpalme! There's a lot of code to review, so I haven't looked thoroughly at everything yet, but I thought I'd get the ball rolling with a few comments based on an initial quick scan of what has changed. See below!
> 
> Please let me know if/when you'd like help with hands-on testing as well! Obviously my ability to test is limited since I do not have access to OverDrive, but with the fake connector, I can at least do some local testing. I just don't want to start that before you think the code is ready to run, so please keep me informed of the status. 

I didn't create the FakeConnector so I didn't pay much attention to it but I'll clean it up now. I've been meaning to ask if I should try to flush it out more and change the return data to be more realistic.

demiankatz commented:
> I didn't create the FakeConnector so I didn't pay much attention to it but I'll clean it up now. I've been meaning to ask if I should try to flush it out more and change the return data to be more realistic.

I created the FakeConnector because I needed to test and refactor some templates, and I needed a way to be able to see them without OverDrive. If you can easily make it more complete and realistic, that would certainly be helpful, especially if you want my assistance with deeper testing. However, I don't want to make extra work for you if you don't think this is a good use of your time, and I can always add fake data on my end based on guesswork if I need to test something -- the results just might not be as good as if you do it with your more informed perspective. :-)

Commit:
fix typos and formatting in config file

Commit:
reduced logging and fixed translation task

Commit:
adding in translations that got missed.

Commit:
removed a lot of log statments, fixed some typos and a suggested code improvement

Commit:
added tab to plugin manager and removed unnec factory class

Commit:
removed obsolete overridden template

Commit:
implemented suggested changes from another PR

Commit:
adding missing call to get user access

Commit:
fix some issues with magazine support and editing hold suspensions

Commit:
fix missing translations for previous commit

Commit:
remove config for showing OD admin page that wasn't working

Commit:
rearrange and dedupe lang file entries

Commit:
fix non-working test

Commit:
fix overdrive help link

Commit:
formatting fixes

Commit:
formatting fixes pt 2

Commit:
formatting fixes pt 3

bpalme commented:
> If you can easily make it more complete and realistic, that would certainly be helpful, especially if you want my assistance with deeper testing. However, I don't want to make extra work for you if you don't think this is a good use of your time, and I can always add fake data on my end based on guesswork if I need to test something -- the results just might not be as good as if you do it with your more informed perspective. :-)

I was planning on capturing some of the actual returned data for a few real titles and putting that into the test but even serialized it was quite large so I didn't think you would want that cluttering up the code. So I focussed more just getting it done and tested manually.

bpalme commented:
For the changes in PR #3244, I've incorporated those here (One template was removed).

demiankatz committed:
Merge branch 'dev' into VUFIND-1499+VUFIND-1375

demiankatz committed:
Fix capitalization of OverDrive.

demiankatz committed:
Remove extra space.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/status-full.phtml:
```diff
@@ -17,26 +17,38 @@ foreach ($ids as $id) {
+        } else {
+            $current['full_status'] .= $avail->copiesAvailable . ' / ' . $avail->copiesOwned;
+        }
+        $current['full_status'] .= ' ' . $this->transEsc('od_copies_available');
```
Rather than concatenating strings here, it would probably be better to change the od_copies_available string to something like `%%copies%% copies available` and then change this code to something like:

```
$copies = $avail->availabilityType == 'AlwaysAvailable'
    ? $this->transEsc('od_unlimited');
    : ($avail->copiesAvailable . ' / ' . $avail->copiesOwned);
$current['full_status'] .= ' ' . $this->transEsc('od_copies_available', ['%%copies%%' => $copies]);
```

...and then languages that put things in different orders can adjust things appropriately in translation.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/status-full.phtml:
```diff
@@ -17,26 +17,38 @@ foreach ($ids as $id) {
     if ($noshow) {
-        $current['full_status'] = '';
-    } elseif ($isError || $avail->code == 'od_code_resource_not_found') {
+        //patron does not have access to overdrive. don't show avail
```
Minor style: please put a space after `//` in comments.

demiankatz commented on themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml:
```diff
@@ -0,0 +1,182 @@
+        <?php endif; ?>
+        <?= $this->transEsc('od_copies_available') ?>
+        <?php if ($avail->data->numberOfHolds > 0): ?>
+          <?= '; ' . $avail->data->numberOfHolds . ' ' . ($avail->data->numberOfHolds > 1 ? $this->transEsc('od_people') : $this->transEsc('od_person')) . ' ' . $this->transEsc('od_waiting') ?>
```
This looks like an opportunity to use our recently-added support for ICU message strings. Instead of separate od_people/od_person strings, I'd suggest changing this to:

```
<?= '; ' . $this->transEsc('od_waiting', ['holds' => $avail->data->numberOfHolds], null, true) ?>
```

And then you can make the od_waiting translation string:

```
od_waiting = "{holds, plural, =1 {1 person} other {# people}} waiting"
```

This gives more fine-grained control over pluralization support in languages with more complex requirements.

demiankatz commented on languages/en.ini:
```diff
@@ -928,63 +928,92 @@ number_decimal_point = "."
-od_video-streaming = "streaming video file"
+od_susp_after = "Deliver after %%days%% days"
+od_susp_asap = "Deliver as soon as possible"
+od_susp_deliver = "Deliver"
```
The `od_susp_deliver` string does not appear to be used. Is this intentional?

demiankatz committed:
Whitespace cleanup.

demiankatz committed:
Fix incomplete comment.

demiankatz committed:
Add end-of-file whitespace.

demiankatz committed:
Add missing translation; fix trailing whitespace.

demiankatz commented on module/VuFind/src/VuFind/RecordTab/OverdriveHoldings.php:
```diff
@@ -0,0 +1,54 @@
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:record_tabs Wiki
+ */
+class OverdriveHoldings extends AbstractBase implements TabInterface
```
You don't need the `implements TabInterface` here since that is inherited from AbstractBase automatically.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/previewlink.phtml:
```diff
@@ -0,0 +1,5 @@
+<?php
+$previews = $this->driver->getPreviewLinks();
+if ($previews->url): ?>
+    <a href="<?=$previews->url?>>" target="_blank" class="btn btn-info btn-xs"><?=$this->transEsc('Preview')?> </a>
```
Should there be a `$this->escapeHtmlAttr()` around `$previews->url`? Is the blank space really wanted before the ` </a>`?

demiankatz reviewed (changes requested):
Thanks for the latest progress, @bpalme! I haven't had a chance to exhaustively review everything yet, but I spent some time today doing a few things:

1.) I merged the latest dev branch and resolved a few minor conflicts.
2.) I updated all of the language files to use the correct "OverDrive" capitalization instead of "Overdrive."
3.) I fixed a couple of small typos.
4.) I left a few comments (see below), mostly focused on internationalization.

I may do another review and provide further feedback on other files later if time permits. In the meantime, please let me know if you have any questions about, or need any help with, this current round of feedback.

Commit:
merge dev into VUFIND-1499+VUFIND-1375

Commit:
translation improvements and a couple formatting fixes

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/status-full.phtml:
```diff
@@ -17,26 +17,36 @@ foreach ($ids as $id) {
+            $current['full_status'] .= '</span> ';
+        }
+        $copies = $avail->availabilityType == 'AlwaysAvailable'
+            ? $this->transEsc('od_unlimited');
```
The semi-colon at the end of this line should be removed; it's causing a syntax error.

demiankatz reviewed (changes requested):
Thanks for the ongoing progress, @bpalme -- I just happened to notice there was a continuous integration failure here, and it appears to be caused by the minor issue below.

Commit:
merge dev into VUFIND-1499+VUFIND-1375

Commit:
remove pesky semicolon

demiankatz commented on themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml:
```diff
@@ -0,0 +1,182 @@
+          </span>
+        <?php endif; ?>
+
+        <?php if (isset($avail->data->availabilityType) && $avail->data->availabilityType == 'AlwaysAvailable'): ?>
```
Looks like this code needs to be revised to match the edits in status-full.phtml related to od_copies_available. Also, in en.ini, it looks like od_copies_available needs to have the `%%copies%%` token added to the front of it.

demiankatz replied:
It looks like this comment may have gotten missed...

bpalme replied:
Either I misunderstood what you meant or the system just didn't pick up the change because the line number is little different:  https://github.com/bpalme/vufind/blob/VUFIND-1499%2BVUFIND-1375/themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml#L48

demiankatz replied:
Strange -- when I looked at this yesterday, `od_copies_available` was still missing the `%%copies%%` token in en.ini. But I see it's there now! Sometimes GitHub lags behind, and changes don't show up right away. Perhaps that is what happened. In any case, thanks for all the progress, and I'll review this again as soon as I can (probably early next week, but maybe sooner if I'm lucky ;-) ).

demiankatz commented on module/VuFind/src/VuFind/Controller/OverdriveController.php:
```diff
@@ -99,11 +99,38 @@ public function mycontentAction()
                 $checkoutsUnavailable = true;
             } else {
                 foreach ($checkoutResults->data as $checkout) {
+                    $mycheckout = [];
```
You could shave off a line of code by replacing these two lines with:

`$mycheckout = compact('checkout');`

That will initialize $mycheckout to ['checkout' => $checkout] for you.

demiankatz commented on module/VuFind/src/VuFind/Controller/OverdriveController.php:
```diff
@@ -258,41 +333,114 @@ public function holdAction()
+                $result->code = 'od_hold_update_failure';
+            }
+        } elseif ($action == 'editSuspendedHold') {
+            $actionTitleCode = 'od_susp_hold_edit';
```
This method is very long; it might be easier to read, troubleshoot and extend if it could be broken down into smaller parts. I certainly don't insist that you do this right now, but it might be helpful. For example, might it be possible to break down this long if/else statement into something like this?

```
$actions = [
    'editSuspendedHold' => ['titleCode' => 'od_susp_hold_edit', 'resultsMethod' => 'getEditSuspendedCodeResult'],
    // other action definitions go here...
];
if (isset($actions[$action])) {
    $actionTitleCode = $actions[$action]['titleCode'];
    $result = ($this->{$actions[$action]['resultsMethod']})();
} else {
    // action not defined logic goes here
}
```

And then you end up with a whole lot of small action processing methods instead of one big one.

Maybe there are reasons it's not as simple as this, but just an idea. If you like the idea but don't have time to deal with this right now, I'm happy to get this merged as-is and can help with refactoring in the future, as long as you can help with testing. :-)

bpalme replied:
I like the idea. I've been meaning to but just haven't gotten up the gumption yet. I'll start working on it.

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -263,7 +262,7 @@ public function getAvailability($overDriveId)
             $res = $this->callUrl($availabilityUrl);
 
-            if ($res->errorCode == 'NotFound') {
+            if (isset($res->errorCode) and $res->errorCode == 'NotFound') {
```
You can simplify some conditions in this class by using null coalescing instead of isset checks. For example, this could become:

`if (($res->errorCode ?? '') == 'NotFound') {`

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -1037,34 +1242,50 @@ public function getHolds($refresh = true)
-                                = $holdExpires->format(
+
+                    if (isset($response->holds)) {
+                        $result->data = [];
```
You're assigning values to `$result->data` in two consecutive lines... since the empty array assigned in the first line is immediately overwritten, you don't need the first line at all.

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -1037,34 +1242,50 @@ public function getHolds($refresh = true)
+                        // Check for holds ready for chechout
+                        foreach ($response->holds as $key => $hold) {
+                            // check for hold suspension
+                            if (isset($hold->holdSupension)) {
```
This if/else could be simplified to:

`$result->data[$key]->holdSupension = $hold->holdSupension ?? false;`

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -1450,12 +1673,13 @@ protected function connectToPatronAPI(
+     * @param bool   $allowRedirects Whether to allow the client to follow redirects
      *
-     * @return \Laminas\Http\Client
+     * @return \Zend\Http\Client
```
This Zend should be changed back to Laminas; probably reintroduced by accident.

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php:
```diff
@@ -250,8 +250,7 @@ public function getPreviewLinks()
     {
-        // TODO: add this as an Overdrive configuration to turn it off
-        return true;
+        return $this->config->enableAjaxStatus;
```
Should we have a default value here in case enableAjaxStatus is unset? e.g. `return $this->config->enableAjaxStatus ?? true;`?

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/hold.phtml:
```diff
@@ -1,59 +1,81 @@
+  <?php if($action == 'getTitle'): ?>
+    <h2><?=$this->transEsc($actionTitleCode); ?>
+        <span class="odbrand pull-left">
+          <img class="media-left" src="https://developerportaldev.blob.core.windows.net/media/Default/images/newLogos/OverDrive_Logo_28x28_rgb.jpg">
```
Would it be better to include a local copy of this file rather than linking to it externally, or does OverDrive require this approach? The external links may get blocked by content security policies, plus it makes us dependent on external hosting that might change over time.

demiankatz commented on themes/bootstrap3/templates/admin/menu.phtml:
```diff
@@ -5,7 +5,5 @@
-  <?php if ($this->overdrive()->showOverdriveAdminLink()):?>
-    <a href="<?=$this->url('admin/overdrive')?>" class="list-group-item<?=strtolower($this->layout()->templateName) == 'overdrive' ? ' active' : ''?>"><?=$this->transEsc('Overdrive API')?></a>
-  <?php endif;?>
+  <a href="<?=$this->url('admin/overdrive')?>" class="list-group-item<?=strtolower($this->layout()->templateName) == 'overdrive' ? ' active' : ''?>"><?=$this->transEsc('Overdrive API')?></a>
```
I don't think the string Overdrive API exists in en.ini -- and shouldn't it be OverDrive API in any case?

demiankatz commented on themes/bootstrap3/templates/myresearch/odmycontent.phtml:
```diff
@@ -55,6 +68,8 @@
             </a>
           </h4>
 
+          <div class="od_edition"><?=$edition?></div>
```
Does $edition need to be escaped here, or do we expect it to contain HTML?

demiankatz commented on themes/bootstrap3/templates/myresearch/odmycontent.phtml:
```diff
@@ -64,23 +79,23 @@
+                class="btn btn-warning icon-link returnTitle"
                 data-lightbox title="<?=$this->transEsc('od_but_return')?>"
-                href="<?=$this->url('overdrive-hold') . "?od_id=$od_id&rec_id=$rec_id&action=returnTitleConfirm"; ?>"
+                href="<?=$this->url('overdrive-hold') . "?od_id=$od_id&rec_id=$rec_id&edition=$edition&action=returnTitleConfirm"; ?>"
```
Does $edition need to be urlencoded here?

demiankatz reviewed (changes requested):
Thanks for the ongoing progress, @bpalme -- I took another pass through the files and found a few more things to comment on. I hope this is helpful! Please let me know when you're ready for me to do more review and testing.

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php:
```diff
@@ -361,13 +374,15 @@ public function isCheckedOut()
-            foreach ($holds as $hold) {
-                if (strtolower($hold->reserveId) == $overDriveId) {
-                    return $hold;
+        if ($this->isLoggedIn()) {
```
In #3459, I eliminated the isLoggedIn method that returns a user object, since it's a bit confusing to have a method that sounds like a Boolean but returns a non-Boolean value. While working on that, I noticed that Overdrive had an isLoggedIn method of its own, but that method is not used at all in the dev branch. I see that you _are_ using it here. I'd suggest either renaming the method to something like `getCatalogUser` if you want it to return an object, or changing it to return a Boolean value (since that seems to be the only way you're using it anyway).

bpalme replied:
I changed the return type. But let me know if there is a better way to do it. I couldn't find any other drivers that checks for a user so maybe there is a more fundamental problem here.

demiankatz replied:
I think this is fine for now at least!

Commit:
refactored controller action along with some other suggested fixes

Commit:
merge dev into VUFIND-1499+VUFIND-1375

Commit:
change return type of isLoggedIn and fix unresolved merge

Commit:
fix some formatting issues

Commit:
formatting fix

Commit:
formatting fix2

Commit:
remove duplicate key in array

bpalme commented:
It occurred to me that perhaps it was a faux-pas to resolve the conversations. Is it understood that you are the one to do that? At any rate, I left the others unresolved, but I think I'm all caught up on changes so far.

demiankatz commented:
Thanks, @bpalme -- it looks like one comment related to i18n may have been missed. I've put a follow-up comment on that thread. I'll review the rest of the changes soon, as time permits!

Regarding resolving comment threads, I'm happy with whatever workflow works best for you. If you leave a comment and you want me to see/acknowledge it, it's probably best to leave the thread open -- then I can read what you said, double-check the changes, and resolve the thread. But if the comment is something really simple, like "fix this typo," then you can fix it and resolve the comment and nothing more need be said. :-)

In any case, thanks for all the work on this -- it's looking like release 10.0 will be coming in May or June, so hopefully we're on track to get this all wrapped up by then.

Commit:
remove unneeded comments

demiankatz committed:
Merge branch 'dev' into VUFIND-1499+VUFIND-1375

demiankatz commented on module/VuFind/src/VuFind/Controller/OverdriveController.php:
```diff
@@ -183,17 +218,23 @@ public function holdAction()
             return $patron;
         }
-        $this->debug('patron: ' . print_r($patron, true));
+        //$this->debug('patron: ' . print_r($patron, true));
```
Can this be fully removed, or might you need it back?

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php:
```diff
@@ -272,11 +271,11 @@ public function getOverdriveAccess()
      * Returns whether the current user is logged in
      *
-     * @return object|bool User if logged in, false if not.
+     * @return bool User if logged in, false if not.
```
This comment is no longer accurate with the changed return type -- I'd just drop the comment entirely since the method is now self-explanatory, and reduce the line to simply `@return bool`

demiankatz committed:
Fix typo.

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php:
```diff
@@ -326,29 +324,44 @@ public function getOverdriveAvailability()
+            if ($result->status) {
+                $checkedout = false;
+                $checkouts = $result->data;
+                //In case of a magazine issue, we have to get all the checkouts to see if the
```
Minor style nitpick: we generally include a space after `//` comments.

demiankatz committed:
Minor typo fix.

demiankatz reviewed (changes requested):
@bpalme, I had a little bit of time to look at this again, so I resolved all of the open issues that are taken care of, and left a few more very minor suggestions. I'll try to find time to give the remaining files a more thorough review soon!

Also note that I merged the latest dev branch into this one and corrected a couple of very minor issues here to save you some time.

Commit:
fix to prevent OD connector from contacting API for permanent link

Commit:
fix comment

Commit:
Merge branch 'VUFIND-1499+VUFIND-1375' of https://github.com/bpalme/vufind into VUFIND-1499+VUFIND-1375

Commit:
rm unnec comment

Commit:
rm unwanted whitespace

Commit:
change array definition

Commit:
rm unwanted whitespace

demiankatz committed:
Fix typo.

demiankatz committed:
Simplify comment/

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php:
```diff
@@ -549,4 +562,39 @@ public function getRealTimeTitleHold()
+     */
+    public function getPermanentLink()
+    {
+        $libraryURL = $this->config->libraryURL;
```
Where does this libraryURL setting come from? I don't see it in Overdrive.ini. Am I looking in the wrong place, or do we need to add an example? Also, do we need to handle the possibility that the setting is missing? Should we throw an exception in that case, or just skip providing a URL?

bpalme replied:
Strange. yes it was missing before and I am sure I've added it at least once. I guess it somehow didn't get committed. I think we do need the URL since it's library specific. I'll re-add it to config.

demiankatz reviewed (changes requested):
Thanks, @bpalme -- I'm still chipping away at reviewing this. It looks like all open comments from prior reviews are now addressed (there were one or two small things that seem to have gotten missed, but I just went ahead and fixed them because they were minor and didn't need to waste more of your time). See below for one new question. I'll continue reviewing as time permits. My goal is to get this merged within the next 2-3 weeks so we can get all of the language file updates in place in time for the 10.0 release... but if that timeline doesn't work for you, we can always push this forward to a future release again.

demiankatz commented on themes/bootstrap3/templates/admin/menu.phtml:
```diff
@@ -5,7 +5,5 @@
-  <?php if ($this->overdrive()->showOverdriveAdminLink()):?>
-    <a href="<?=$this->url('admin/overdrive')?>" class="list-group-item<?=strtolower($this->layout()->templateName) == 'overdrive' ? ' active' : ''?>"><?=$this->transEsc('Overdrive API')?></a>
-  <?php endif;?>
+  <a href="<?=$this->url('admin/overdrive')?>" class="list-group-item<?=strtolower($this->layout()->templateName) == 'overdrive' ? ' active' : ''?>"><?=$this->transEsc('od_admin_menu')?></a>
```
Is there a reason why we made the Overdrive menu item unconditional here? Might we still want the ability to disable it when it's not relevant?

bpalme replied:
Well the permission wasn't working for some reason and I noticed none of the others had something like that so I changed it to be like them. If Overdrive is not configured, it just displays empty values which seems appropriate to me. But let me know if you want me to put it back and make it work again.

bpalme replied:
This is probably why it wasn't working: [VUFIND-1398](https://openlibraryfoundation.atlassian.net/browse/VUFIND-1398)

demiankatz replied:
I took a couple of minutes to look into this and pushed up commit 3a3bd264fe51d6b65e069a808b433e53e7837a89 to reinstate the configuration. I think it's worth having, since people may have varying levels of comfort with providing web-based access to data about their settings, etc. Of course, feel free to adjust if you don't like anything I've done here.

demiankatz commented on themes/bootstrap3/templates/myresearch/odmycontent.phtml:
```diff
@@ -55,6 +68,8 @@
             </a>
           </h4>
 
+          <div class="od_edition"><?=$this->escapeHtml($edition)?></div>
```
Do we want to display this div even if $edition is an empty string? Would it make sense to wrap in `<?php if ($edition): ?> ... <?php endif; ?>`?

demiankatz commented on themes/bootstrap3/templates/myresearch/odmycontent.phtml:
```diff
@@ -64,23 +79,24 @@
+                class="btn btn-warning icon-link returnTitle"
                 data-lightbox title="<?=$this->transEsc('od_but_return')?>"
-                href="<?=$this->url('overdrive-hold') . "?od_id=$od_id&rec_id=$rec_id&action=returnTitleConfirm"; ?>"
+                href="<?=$this->url('overdrive-hold') . "?od_id=$od_id&rec_id=$rec_id&edition=" .
```
Are the various IDs here guaranteed numeric, or should we be urlencoding all of them?

bpalme replied:
As far as I know they are always alphnum but I'll just encode them to be on the safe side.

demiankatz commented on module/VuFind/src/VuFind/Controller/OverdriveController.php:
```diff
@@ -121,16 +146,27 @@ public function mycontentAction()
+                            ->load(strtolower($hold->reserveId));
+                        $holds[] = $myhold;
+                    } catch (\VuFind\Exception\RecordMissing $e) {
+                        //hold is missing from Solr
```
Nitpick: I'd add a space after `//` here and below on line 162.

demiankatz commented on module/VuFind/src/VuFind/Controller/OverdriveController.php:
```diff
@@ -210,90 +249,36 @@ public function holdAction()
+
+        if (isset($actions[$action])) {
+            $actionTitleCode = $actions[$action]['titleCode'];
+            $result = $actions[$action]['resMeth'] ? $this->{$actions[$action]['resMeth']}() : false;
```
Does it make any difference that in your old code, $result defaulted to `null` but now it gets set to `false` instead for situations where no method is used? I think it's likely the two are functionally equivalent for the purposes of your code, but maybe `null` is a more appropriate value for this situation.

demiankatz commented on module/VuFind/src/VuFind/Controller/OverdriveController.php:
```diff
@@ -308,11 +293,290 @@ public function holdAction()
+        );
+        $formats = $driver->getDigitalFormats();
+        // Looks like this is a magazine...
+        if (current($formats)->id == 'magazine-overdrive') {
```
Is this a safe enough check? Is there always going to be exactly one digital format? If not, would it be better to create a support method to search the whole array?

bpalme replied:
I think it is. Ebooks and audiobooks can have multiple formats but magazines just come in one format

demiankatz commented on module/VuFind/src/VuFind/Controller/OverdriveController.php:
```diff
@@ -308,11 +293,290 @@ public function holdAction()
+            $rec_id
+        );
+        $formats = $driver->getDigitalFormats();
+        $result->data = (current($formats)->id == 'magazine-overdrive') ?: false;
```
Here's another instance of the magazine check. Maybe it's worth creating a support method like `isRecordDriverMagazine` to do this work since it's repeated, whether or not the logic itself needs to be adjusted.

demiankatz commented on themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml:
```diff
@@ -0,0 +1,182 @@
+  <div class="internetDetails">
+    <div class="odbrand pull-left media-left">
+      <img
+        src=""<?=$this->imageLink('OverDrive_Logo_28x28_rgb.jpg')?>"">
```
There seem to be extra double-quotes here.

demiankatz commented on themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml:
```diff
@@ -0,0 +1,182 @@
+      </p>
+
+      <?php if (!$checkedOut || $checkedOutResult->isMagazine): ?>
+        <?php if (isset($avail->data->copiesAvailable) && $avail->data->copiesAvailable > 0): ?>
```
You could simplify some of the if checks in this file by removing issets through null coalescing -- e.g.:

` <?php if (isset($avail->data->copiesAvailable) && $avail->data->copiesAvailable > 0): ?>`

could become:

` <?php if (($avail->data->copiesAvailable ?? 0) > 0): ?>`

No pressure to do this if it's too much work or you prefer not to, but making the statements more concise might be helpful.

demiankatz commented on themes/bootstrap3/templates/RecordTab/overdriveholdings.phtml:
```diff
@@ -0,0 +1,182 @@
+        <?php if (isset($avail->data->copiesAvailable) && $avail->data->copiesAvailable > 0): ?>
+          <p>
+            <a class="btn btn-primary checkout" data-lightbox title="<?= $this->transEsc('request_place_text') ?>"
+              href="<?= "$hold_url?od_id=$od_id&rec_id=$rec_id&action=checkoutConfirm" ?>">
```
I think I've asked this question elsewhere, but this is another place where I wonder if we need to urlencode some of these IDs just in case they contain unexpected characters. But maybe it's not necessary if they're absolutely guaranteed to be integers.

demiankatz reviewed (changes requested):
Thanks again, @bpalme -- I managed to look through a couple more files; see a few more suggestions below. Your refactoring of the controller is a big improvement over the earlier version, in any case!

bpalme commented:
> Thanks, @bpalme -- I'm still chipping away at reviewing this. It looks like all open comments from prior reviews are now addressed (there were one or two small things that seem to have gotten missed, but I just went ahead and fixed them because they were minor and didn't need to waste more of your time). See below for one new question. I'll continue reviewing as time permits. My goal is to get this merged within the next 2-3 weeks so we can get all of the language file updates in place in time for the 10.0 release... but if that timeline doesn't work for you, we can always push this forward to a future release again.

I would very much like to get this in the next release. I should have the current batch of revisions done by this Saturday.

demiankatz commented:
Excellent, thanks, @bpalme. I have just two long files left to review, so I'll see if I can get to those this week as well just in case there are more comments for me to share. Hopefully nothing major in any case!

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/hold.phtml:
```diff
@@ -1,59 +1,81 @@
+                <?php endforeach;?>
+                </select>
             </div>
+            <?php  endif;?>
```
Minor style thing -- might be worth a string replace from `<?php  ` (2 spaces) to `<?php ` (1 space) to clean up some unnecessary whitespace in a couple of spots here. Also worth adding a space before the `?>` in this close tag.

bpalme replied:
I think I caught all of these. Also removed all space after short echo tags based on other templates I looked at.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/hold.phtml:
```diff
@@ -1,59 +1,81 @@
+          <img class="media-left" src="<?=$this->imageLink('OverDrive_Logo_28x28_rgb.jpg')?>">
+        </span>
+    </h2>
+    <iframe class="od_download" src="<?=$result->data->downloadRedirect?>"></iframe>
```
Might be worth `$this->escapeHtmlAttr()` around the src value here.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/hold.phtml:
```diff
@@ -1,59 +1,81 @@
+            <div class="alert alert-warning">
+                <div class="moreInfo">
+                    <?=$this->transEsc('od_docheckout_failure') ?>
+                    <?=$result->msg?>
```
This might need `$this->escapeHtml` if it's not already in HTML.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/hold.phtml:
```diff
@@ -116,144 +134,140 @@
+        <form class="" method="POST" name="editHoldEmail" data-lightbox-onclose="VuFind.refreshPage">
+            <div class="form-group">
+                <label class="control-label"><?=$this->transEsc('Email Address'); ?></label>
+                <input id="email" name="email" value="<?=$result->data->hold->emailAddress?>" size="" class="form-control" type="text">
```
`$this->escapeHtmlAttr` on the email address?

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/hold.phtml:
```diff
@@ -116,144 +134,140 @@
+                <label id="select-suspension-label"><?=$this->transEsc('od_susp_hold_confirm'); ?></label>
+                <select class="form-control" name="suspendValue" id="select-suspension" data-open="false">
+                    <option value="0" selected=""><?=$this->transEsc('od_susp_asap'); ?></option>
+                    <option value="7"><?=$this->transEsc('od_susp_after', ['%%days%%' => '7']); ?></option>
```
Would it make sense to refactor this to:

```
<?php $durations = [7, 14, 21, 28, 60, 90, 180]; ?>
<?php foreach ($durations as $duration): ?>
  <option value="<?=$duration?>"><?=$this->transEsc('od_susp_after', ['%%days%%' => $duration]); ?></option>
<?php endforeach; ?>
```

?

That would also open the door to making this a config setting if there were any value in doing that.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/hold.phtml:
```diff
@@ -116,144 +134,140 @@
-                    <?php endforeach;?>
-                </div>
+                    <option value="7"><?=$this->transEsc('od_susp_after', ['%%days%%' => '7']); ?></option>
+                    <option value="14"><?=$this->transEsc('od_susp_after', ['%%days%%' => '14']); ?></option>
```
Ditto here.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/hold.phtml:
```diff
@@ -116,144 +134,140 @@
+    <div class="returnTitleDialog">
+
+        <?php  if ($result->data): ?>
+          <h4><?=$edition?></h4>
```
Does this need to be escaped?

demiankatz commented on module/VuFind/src/VuFind/DigitalContent/OverdriveConnector.php:
```diff
@@ -326,7 +330,7 @@ public function getAvailabilityBulk($overDriveIds = [])
                 $result->code = 'od_code_connection_failed';
             } else {
-                if ($res->errorCode == 'NotFound' || $res->totalItems == 0) {
+                if ((isset($res->errorCode) and $res->errorCode == 'NotFound') or $res->totalItems == 0) {
```
Could this be:

```
if (($res->errorCode ??  null) == 'NotFound' || $res->totalItems == 0) {
```

demiankatz reviewed (changes requested):
Okay, I think I've finished looking at everything -- here are a few more questions and nitpicks, but hopefully nothing that will take up too much of your time!

Commit:
merge dev into PR branch

Commit:
missing config item

Commit:
added configuration for hold suspensions, more escaping in templates, fix formatting

Commit:
formatting fixes

Commit:
replacing double quotes with singles

Commit:
template formatting

demiankatz committed:
Fix/reinstate showOverdriveAdminMenu setting.

demiankatz committed:
Fix case of OverDrive in comments.

demiankatz committed:
Minimize diffs.

demiankatz committed:
Fix case.

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/SolrOverdrive.php:
```diff
@@ -549,4 +562,42 @@ public function getRealTimeTitleHold()
+    /**
+     * Get Permanent Link to the resource on your institution's OverDrive site
+     *
+     * @return string the permanent link to the resource
```
Looks like the return type here is actually `array` rather than `string`.

demiankatz committed:
Code simplification.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/hold.phtml:
```diff
@@ -1,59 +1,81 @@
+        </span>
     </h2>
-
+         <div class="pull-right  img-col">
```
There's an extra space between the two classes here.

demiankatz commented on themes/bootstrap3/templates/RecordDriver/SolrOverdrive/hold.phtml:
```diff
@@ -111,149 +129,138 @@
+                <label id="select-suspension-label"><?=$this->transEsc('od_susp_hold_confirm')?></label>
+                <select class="form-control" name="suspendValue" id="select-suspension" data-open="false">
+                    <option value="0" selected=""><?=$this->transEsc('od_susp_asap')?></option>
+                    <?php $durations = $this->config()->get('Overdrive')->Overdrive->suspendDurations; ?>
```
Is it worth having a default fallback in case the config is unset, e.g.

```
<?php $durations = $this->config()->get('Overdrive')->Overdrive->suspendDurations ?? [7, 14, 21, 28, 60, 90, 180]; ?>
```

That way outdated configurations won't break the functionality. If you do it here, you'll also need to do it in the other place that displays the menu.

demiankatz reviewed (changes requested):
Thanks again, @bpalme -- just a couple more very trivial suggestions. I also pushed up a few more minor adjustments while I was reviewing the latest changes. If you clean up these last couple of things, and as long as you don't object to what I've done with the admin code, we may be ready to merge this very soon.

demiankatz committed:
Merge branch 'dev' into VUFIND-1499+VUFIND-1375

Commit:
default config setting for hold durations

Commit:
remove space

Commit:
Merge branch 'VUFIND-1499+VUFIND-1375' of https://github.com/bpalme/vufind into VUFIND-1499+VUFIND-1375

demiankatz committed:
Adjust whitespace.

demiankatz reviewed (approved):
Thanks, @bpalme, for all the hard work on this (and for bearing with all of my feedback). As far as I can tell, this seems to be ready to merge now. Are there any last-minute things you want to do before I hit the merge button?

bpalme commented:
No. I say go for it. I'll try to do another round of testing after you merge and before we do the final release.

bpalme commented:
I have some related edits for the ovedrive wiki page. When should I add those and what is the process?

demiankatz commented:
@bpalme, feel free to make those edits at your convenience -- just note which things are 10.0-specific, if that matters. And if you have trouble accessing the wiki, let me know.
